kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -747,31 +753,108 @@ impl<'tcx> GotocCtx<'tcx> {
-            self.cast_expr_to_unsized_expr(src_goto_expr, src_mir_type, dst_mir_type)
-        }
+    ) -> Expr {
+        // Collect some information about the unsized coercion.
+        let mut path = self.collect_unsized_cast_path(src_goto_expr, src_mir_type, dst_mir_type);
+        debug!(cast=?path, "codegen_unsized_cast");
+
+        // Handle the leaf which should always be a pointer.
+        let (ptr_cast_info, ptr_src_expr) = path.pop().unwrap();
+        let initial_expr = self.codegen_cast_to_fat_pointer(ptr_src_expr, ptr_cast_info);
+
+        // Iterate from the back of the path initializing each struct that requires the coercion.
+        // This code is required for handling smart pointers.
+        path.into_iter().rev().fold(initial_expr, |coercion_expr, (info, src_expr)| {
+            self.codegen_struct_unsized_coercion(src_expr, info, coercion_expr)
+        })
+    }
+
+    /// Extract path that must be explicitly casted. Add to the tuple the expression type for
+    /// the current source.
+    fn collect_unsized_cast_path(
+        &self,
+        src_goto_expr: Expr,
+        src_mir_type: Ty<'tcx>,
+        dst_mir_type: Ty<'tcx>,
+    ) -> Vec<(CoerceUnsizedInfo<'tcx>, Expr)> {
+        let mut field_type = src_goto_expr;
+        CoerceUnsizedIterator::new(self.tcx, src_mir_type, dst_mir_type)
+            .map(|info| {
+                let expr = if let Some(field_symbol) = info.field {
+                    let field_name = field_symbol.as_str().intern();
+                    let member_type = field_type.clone().member(field_name, &self.symbol_table);
+                    std::mem::replace(&mut field_type, member_type)
+                } else {
+                    field_type.clone()
+                };
+                (info, expr)
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Generate a struct resulting from an unsized coercion.
+    /// The resulting expression is basically a field by field assignment to from the
+    /// source expression, except for the field being coerced.
+    /// Coercion ignores phantom data structures, so do we.
+    /// See <https://github.com/rust-lang/rust/issues/26905> for more details.
+    fn codegen_struct_unsized_coercion(
+        &mut self,
+        src_expr: Expr,
+        info: CoerceUnsizedInfo<'tcx>,
+        member_coercion: Expr,
+    ) -> Expr {
+        assert!(info.src_ty.is_adt(), "Expected struct. Found {:?}", info.src_ty);
+        assert!(info.dst_ty.is_adt(), "Expected struct. Found {:?}", info.dst_ty);
+        let dst_goto_type = self.codegen_ty(info.dst_ty);
+        let src_field_exprs = src_expr.struct_field_exprs(&self.symbol_table);
+        let dst_field_exprs = src_field_exprs
+            .into_iter()
+            .map(|(key, val)| {
+                let new_val = if info.field.unwrap().as_str().intern() == key {
+                    // The type being coerced. Use the provided expression.
+                    member_coercion.clone()
```

Comment:
Can we avoid the clone? Not a big deal...

Reply:
I don't think we can because it would attempt to move the captured value.

Reply:
Ohhh, right.  WE know this only happens once, but the borrow checker sees a function that gets run repeatedly. Ok.